contracts/src/UniswapV3MultiQuoter.sol

```diff
@@ -0,0 +1,301 @@
+            (state.sqrtPriceX96, step.amountIn, step.amountOut, step.feeAmount) = SwapMath.computeSwapStep(
+                state.sqrtPriceX96,
+                (zeroForOne ? step.sqrtPriceNextX96 < sqrtPriceLimitX96 : step.sqrtPriceNextX96 > sqrtPriceLimitX96)
+                    ? sqrtPriceLimitX96
+                    : step.sqrtPriceNextX96,
+                state.liquidity,
+                state.amountSpecifiedRemaining,
+                fee
+            );
+
+            if (exactInput) {
+                state.amountSpecifiedRemaining -= (step.amountIn + step.feeAmount).toInt256();
+                state.amountCalculated = state.amountCalculated.sub(step.amountOut.toInt256());
+            } else {
+                state.amountSpecifiedRemaining += step.amountOut.toInt256();
+                state.amountCalculated = state.amountCalculated.add((step.amountIn + step.feeAmount).toInt256());
+            }
+
+            // shift tick if we reached the next price
+            if (state.sqrtPriceX96 == step.sqrtPriceNextX96) {
+                // if the tick is initialized, run the tick transition
+                if (step.initialized) {
+                    (, int128 liquidityNet, , , , , , ) = pool.ticks(step.tickNext);
+
+                    // if we're moving leftward, we interpret liquidityNet as the opposite sign
+                    // safe because liquidityNet cannot be type(int128).min
+                    if (zeroForOne) liquidityNet = -liquidityNet;
+
+                    state.liquidity = LiquidityMath.addDelta(state.liquidity, liquidityNet);
+                }
+
+                state.tick = zeroForOne ? step.tickNext - 1 : step.tickNext;
+            } else if (state.sqrtPriceX96 != step.sqrtPriceStartX96) {
+                // recompute unless we're on a lower tick boundary (i.e. already transitioned ticks), and haven't moved
+                state.tick = TickMath.getTickAtSqrtRatio(state.sqrtPriceX96);
+            }
+
+            if (state.amountSpecifiedRemaining == 0) {
+                (result.amounts0[state.amountsIndex], result.amounts1[state.amountsIndex]) = zeroForOne == exactInput
+                    ? (amounts[state.amountsIndex], state.amountCalculated)
+                    : (state.amountCalculated, amounts[state.amountsIndex]);
+
+                result.gasEstimates[state.amountsIndex] = gasBefore - gasleft();
+
+                if (state.amountsIndex == amounts.length - 1) {
+                    return (result);
+                }
+
+                state.amountsIndex += 1;
+                state.amountSpecifiedRemaining = amounts[state.amountsIndex].sub(amounts[state.amountsIndex - 1]);
+            }
+        }
+
+        for (uint256 i = state.amountsIndex; i < amounts.length; ++i) {
+            (result.amounts0[i], result.amounts1[i]) = zeroForOne == exactInput
+                ? (amounts[i] - state.amountSpecifiedRemaining, state.amountCalculated)
```

Comment:
would the router logic get confused by this? i.e. returning prorated input amounts if we run our of liquidity instead of returning the original amountsIn

Reply:
I don't believe so. I believe router would only get confused if the output amounts are non-monotonic. In this case the sampler amounts are still monotonically increasing.